core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleGetRecoveryCandidatesHandler.java

```diff
@@ -0,0 +1,187 @@
+                            .map(CloudInstance::getInstanceId)
+                            .collect(Collectors.toSet()));
+
+            ClusterHealthService clusterHealthService = clusterApiConnectors.getConnector(stack).clusterHealthService();
+
+            DetailedHostStatuses detailedHostStatuses =
+                    clusterHealthService.getDetailedHostStatuses();
+
+            flowMessageService.fireEventAndLog(stack.getId(), UPDATE_IN_PROGRESS.name(), CLUSTER_SCALING_STOPSTART_DOWNSCALE_IDENTIFYRECOVERYCANDIDATES,
+                    request.getHostGroupName());
+
+            if (clusterHealthService.isClusterManagerRunning()) {
+                startedInstanceMetadataWithServicesNotRunning = collectRecoveryCandidates(startedInstancesMetadata,
+                        detailedHostStatuses, runtimeVersionService.getRuntimeVersion(stack.getClusterId()));
+            }
+
+            List<CloudInstance> instancesWithServicesNotRunning;
+
+            if (startedInstanceMetadataWithServicesNotRunning.isEmpty()) {
+                instancesWithServicesNotRunning = Collections.emptyList();
+            } else {
+                instancesWithServicesNotRunning = instanceMetaDataToCloudInstanceConverter.convert(startedInstanceMetadataWithServicesNotRunning,
+                        stack.getStack());
+                LOGGER.info("Collected {} downscale recovery candidates for hostGroup: {}", instancesWithServicesNotRunning.size(), request.getHostGroupName());
+            }
+
+            StopStartDownscaleGetRecoveryCandidatesResult result = new StopStartDownscaleGetRecoveryCandidatesResult(stack.getId(),
+                    instancesWithServicesNotRunning, request.getHostGroupName(), request.getHostIds());
+            notify(result, event);
+        } catch (Exception e) {
+            String errorMessage = String.format("Failed to collect downscale recovery candidates for stack: %s. " +
+                    "Proceeding with subsequent flow.", request.getResourceId());
+            LOGGER.warn(errorMessage, e);
+            StopStartDownscaleGetRecoveryCandidatesResult result = new StopStartDownscaleGetRecoveryCandidatesResult(request.getResourceId(),
+                    Collections.emptyList(), request.getHostGroupName(), request.getHostIds());
+            notify(result, event);
+        }
+    }
+
+    private List<CloudInstance> collectStartedInstancesFromCloudProvider(CloudConnector connector, AuthenticatedContext ac,
+            List<CloudInstance> cloudInstances) {
+        List<CloudVmInstanceStatus> vmInstanceStatuses = connector.instances().checkWithoutRetry(ac, cloudInstances);
+        return vmInstanceStatuses.stream()
+                .filter(vm -> InstanceStatus.STARTED.equals(vm.getStatus()))
+                .map(CloudVmInstanceStatus::getCloudInstance)
+                .collect(Collectors.toList());
+    }
+
+    //CHECKSTYLE:OFF
```

Comment:
Added these checkstyle workarounds because this method results in cyclomatic complexity (possibly due to the large no. of booleans)

Reply:
Could you please change the conditions? I think this checkstyle ignore is unnecessary.